ai/cohere/v0/client.go

```diff
@@ -0,0 +1,76 @@
+package cohere
+
+import (
+	"context"
+	"sync"
+
+	cohereSDK "github.com/cohere-ai/cohere-go/v2"
+	cohereClientSDK "github.com/cohere-ai/cohere-go/v2/client"
+	"github.com/cohere-ai/cohere-go/v2/core"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+type cohereClient struct {
+	sdkClient cohereClientInterface
+	logger    *zap.Logger
+	lock      sync.Mutex
+}
+
+type cohereClientInterface interface {
+	Chat(ctx context.Context, request *cohereSDK.ChatRequest, opts ...core.RequestOption) (*cohereSDK.NonStreamedChatResponse, error)
+	Embed(ctx context.Context, request *cohereSDK.EmbedRequest, opts ...core.RequestOption) (*cohereSDK.EmbedResponse, error)
+	Rerank(ctx context.Context, request *cohereSDK.RerankRequest, opts ...core.RequestOption) (*cohereSDK.RerankResponse, error)
+}
+
+func newClient(apiKey string, logger *zap.Logger) *cohereClient {
+	client := cohereClientSDK.NewClient(cohereClientSDK.WithToken(apiKey))
+	return &cohereClient{sdkClient: client, logger: logger, lock: sync.Mutex{}}
+}
+
+func (cl *cohereClient) generateEmbedding(request cohereSDK.EmbedRequest) (cohereSDK.EmbedResponse, error) {
+	respPtr, err := cl.sdkClient.Embed(
+		context.TODO(),
+		&request,
+	)
+	if err != nil {
+		panic(err)
+	}
+	resp := cohereSDK.EmbedResponse{
+		EmbeddingsFloats: respPtr.EmbeddingsFloats,
+	}
+	return resp, nil
+}
+
+func (cl *cohereClient) generateTextChat(request cohereSDK.ChatRequest) (cohereSDK.NonStreamedChatResponse, error) {
+	respPtr, err := cl.sdkClient.Chat(
+		context.TODO(),
+		&request,
+	)
+	if err != nil {
+		panic(err)
+	}
+	resp := cohereSDK.NonStreamedChatResponse{
+		Text:         respPtr.Text,
+		GenerationId: respPtr.GenerationId,
+		Citations:    respPtr.Citations,
```

Comment:
Important

Will need to get the input & output tokens.
NonStreamedChatResponse -> Meta -> BilledUnits -> ApiMetaBilledUnits

Reply:
And, could you help us check what are the difference between the tokens in these two struct?

```
BilledUnits *[ApiMetaBilledUnits](https://pkg.go.dev/github.com/cohere-ai/cohere-go/v2@v2.8.2#ApiMetaBilledUnits) `json:"billed_units,omitempty" url:"billed_units,omitempty"`
Tokens      *[ApiMetaTokens](https://pkg.go.dev/github.com/cohere-ai/cohere-go/v2@v2.8.2#ApiMetaTokens)      `json:"tokens,omitempty" url:"tokens,omitempty"`
```

Reply:
Fixed